### ISSUE TITLE
!monitor inconsistent output?

### ISSUE THREAD

**Issue by DarkerSquirrel** [HyperDbg/HyperDbg#91]
Hello Sina

I am trying to trace memory reads from a process to KUSER_SHARED_DATA :
char * KUSER = (char*)0x7ffe0000;

while(true){
  std::cin.get();
  for (int i = 0; i < 10; i++) 
    printf(" %p, %hhX\n", _ReturnAddress(), KUSER[i]);
}

I am trying this command:
!monitor r 0x7ffe0000 0x7ffe0fff pid 0x2824 script  { if( $ip < 0x7FFFFFFFFFF) { printf("RIP=%llx,PID=%d,KUSD_offset=%llx\n",$ip,$pid,$context); } }

The output of HyperDbg seems to be incosistent,it almost never displays 10 accesses.(sometimes 8 , sometimes 12,rarely 10).
All the times some accesses from an address are printed more than 1 time.

Is this something I miss here or is it a bug?

Also , are  you kind enough to let me know when you finish to implement that PhysicalToVirtual() in scripting so I can get context virtual addr of context?

I have 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz 2.80 GHz
Windows 10 Pro 20H2 19042.1348

The output below in screenshot is from an .asm file like this (as you see , one RIP is getting logged twice):
 movsx   r8d, byte [7FFE0000h]
    movsx   r8d, byte [7FFE0001h]
    movsx   r8d, byte [7FFE0002h]
    movsx   r8d, byte [7FFE0003h]
    movsx   r8d, byte [7FFE0004h]  
    movsx   r8d, byte [7FFE0005h]
    movsx   r8d, byte [7FFE0006h]
    movsx   r8d, byte [7FFE0007h]
    movsx   r8d, byte [7FFE0008h]
    movsx   r8d, byte [7FFE0009h]

![image](https://user-images.githubusercontent.com/33521652/145812525-178e0e24-d99e-4565-bd57-3af3cde0cae7.png)

Thanks

**Comment by SinaKarvandi** [comment]
Yep, I can reproduce the multiline showing of messages and that's really weird. 😳
And that's a really weird problem that needs to be investigated.

**Comment by DarkerSquirrel** [comment]
Thanks a lot for your quick response, Sina.

**Comment by DarkerSquirrel** [comment]
Thanks a lot for quick look into this, Sina.

Other things that come to mind related to this:
1. Doing this quick test from a driver (instead of a process) and adjust checked RIP range for that driver triggers this too?
2. Other commands beside !monitor could be affected by it?

**Comment by SinaKarvandi** [comment]
At a first glance, the problem doesn't exist for a simple buffer.

I used the below code :

```
  char *KUSER = (char *)malloc(0x200);

  printf("%x %x", KUSER, KUSER + 0x200);
  int Tmp = NULL;

  while (true) {
    std::cin.get();

    for (int i = 0; i < 10; i++) {
      printf("%X\n", KUSER[i]);
    }
  }
```

![Capture](https://user-images.githubusercontent.com/13383992/145904182-99fc459a-52c7-464b-87d5-4386a0f059bc.PNG)

**Comment by SinaKarvandi** [comment]
Okay, I am concluded that it's more preferable to use **virtual address** instead of **physical address** in the `$context` of the _!monitor_ command.

Please switch to the `dev` branch in this commit (https://github.com/HyperDbg/HyperDbg/commit/6f622bacab3bff1c074c3cadf6f1a85bb8c2d9e4). `$context` is now the virtual address of what was accessed, not physical.

![Capture](https://user-images.githubusercontent.com/13383992/145909942-c7b0b8cb-4bab-4e26-9be9-47b79ec652f7.PNG)

**Comment by SinaKarvandi** [comment]
I also try to understand what's really happening on that special address (0x7ffe0000) that leads to this weird behavior in HyperDbg.

**Comment by DarkerSquirrel** [comment]
Ok , thanks , I will test the dev branch in few hours and let you know.
Just speaking out loud some redundant info that most likely you are aware of, in case someone is reading this thread:

KUSER_SHARED_DATA is a special memory area in a process, in fact there is only one KUSER_SHARED_DATA area in the kernel which is mapped read only in all process spaces.
I tried modifying once some fields of it using a driver and system went kaboom.

I wonder if this has to do with anything related to this bug. In few hours I will do more tests with reading PEB,TEB and other "sensible" mem pages of a process.

**Comment by SinaKarvandi** [comment]
After spending some time debugging this issue, I figured out the root cause of the error, and unfortunately, it's not easy to solve.

The problem arises from the fact that we designed HyperDbg based on a single EPTP. This was a wrong design decision, but we implemented it as it was substantially more straightforward.

The `KUSER_SHARED_DATA` is a shared page that is accessed at a high rate. Whenever an access is made to this page, an EPT Violation is thrown and handled by HyperDbg. Meanwhile, we update the EPTP, unset every interrupt on a single core, and wait for an MTF to be received on the same core. The problem is precisely here. We never halt other cores because of that event. While other events continue their normal execution, they see an updated version of EPTP which doesn't have a page that causes EPT Violation. That's why we lose some of the events.

The reason why we see some events two or three times is that the current core thinks it has access to read the target page, but other cores update the page attributes again, and the current core is forced to re-trigger the same event.

Please visit: https://docs.hyperdbg.org/design/features/vmm-module/design-of-monitor 
for more information about the **!monitor**'s design.

The only solution that came to my mind was using different EPTPs for each core. The design of such a thing needs synchronization between EPTPs in other cores, and it's only possible through a custom APIC implementation and IDT (Interrupt Descriptor Table), which is not yet implemented in HyperDbg.

Adding the above things to the HyperDbg will break many of the semantics that we used in different parts of the debugger, but sure it has to be done as we have seen other issues caused by single EPTP in other parts of the debugger.


I think it's better to leave it for the future versions as it's not a trivial task and we're about to release the first version of HyperDbg in the coming weeks. By the way, the current version of `!monitor` should be fine working on pages that are not shared or if shared, called with a low rate of execution.

Let me know if you have any suggestions for this.

**Comment by DarkerSquirrel** [comment]
Well, as quick dirty fix , what about disabling all cores but one? Would that make the issue go away?

As a longer term fix, I think making a robust !monitor which should display reliably the output is a must,
which should be implemented as soon as the devs time allow it imho(in the next version?)

I am saying this because hyperdbg will mostly be attractive for RE guys which analyse hard targets,
including modern protections which abuse areas like KUSER_SHARED_DATA for hardware licensing parts,
due to the fact that is mapped read only in process space and it cant be overwritten/spoofed like other 
areas as PEB.
Other protections read million of times these special structures (KUSER,PEB,etc) in order to confuse the reverser
and discourage an approach based on exceptions.

So having a !monitor with problems for this kind of cases, makes hyperdbg hardly usable.

Implementing also the cooperative mode with user mode debuggers also in the next imediate version would be
a huge asset in attracting people to HyperDbg.
Making HyperDbg supplement the shortcomings of user space debuggers looks like a winning strategy,at least
for beginning of the project.

Well talking is cheap and coding is hard :-D , I will do my best (as far as I am able) to help improving HyperDbg.
I will give some more thought this EPTP issue and let you know if something comes to mind.

Thanks a lot for your time and detailed explanations

**Comment by SinaKarvandi** [comment]
> Well, as quick dirty fix , what about disabling all cores but one? Would that make the issue go away?

Well, that will fix the issue but brings a new issue of a halted system. This was one of the design principles of HyperDbg to keep other cores running while the target core is processing its event. Halting all cores is not possible. Also, disabling all cores is also not possible. Windows has a lot of watchdog mechanisms to check if all the cores are active or not. Like throwing NMIs, BSOD will occur immediately if we disable other cores. The fact that commands like [instrumentation step-in](https://docs.hyperdbg.org/commands/debugging-commands/i) disable all cores is because the target core is single-stepping, not running.

> As a longer term fix, I think making a robust !monitor which should display reliably the output is a must, which should be implemented as soon as the devs time allow it imho(in the next version?)

Yep, I agree. That needs to be fixed immediately. That's on the immediate to-do list. But, after implementing support to multiple EPTPs, many functionalities should be re-tested because many functionalities depend on EPTP.

> Well talking is cheap and coding is hard :-D , I will do my best (as far as I am able) to help improving HyperDbg.

Thanks a lot for your feedback. You tested HyperDbg in a way that we never used before, and finding these errors will make HyperDbg better and more stable each time.

Please don't limit yourself to just this command and test other functionalities like the !syscall command, EPT Hooks, and Instrumentation step-in command too. I'm pretty sure you find them interesting and useful, and if there is a bug in those functionalities, we'll be happy to solve them.

Thanks again for your valuable feedbacks.

**Comment by SinaKarvandi** [comment]
The problem is now fixed and will be available at v0.6. HyperDbg now supports a multiple EPTP model for its memory layout. Each core has its own EPTP now.

The source code is available at 'multiple-eptp'.
https://github.com/HyperDbg/HyperDbg/tree/multiple-eptp
